### ISSUE TITLE
Support: why the iteration in the operator== algorithm?

### ISSUE THREAD

**Issue by daniel-v** [google/built_collection.dart#24]
I went through the `BuiltList` implementation and I assume the concept is the same for every built collection type. This is only an educated guess. In either case, what I saw in `BuiltList` can be found here: https://github.com/google/built_collection.dart/blob/f91f2307a919f3675f23cff8fef9be8daa9174bf/lib/src/list/built_list.dart#L65-74

Let's assume the following:
We create two `BuiltList<String>` collection instances. Both have got the same strings in it. 

```
var al = new BuiltList<String>([ 'a', 'b' ]);
var bl = new BuiltList<String>([ 'a', 'b' ]);
```

We know that strings are always going to be equal if they are built up of the same code units. It's hashing algorithm will take that into account.

```
var a = 'a'; 
var b = 'a';
print('${a.hashCode == b.hashCode && a == b}'); // true
```

Since `BuiltList` combines the hashCodes of it's members, the hashCode of 2 BuiltList instances are going to be equal if their members have the same hashCodes.

```
var aa = [ 'a', 'b' ];
var bb = [ 'a', 'b' ];
print("${quiver.hashObjects( aa ) == quiver.hashObjects( bb )} "); // true
```

It stands to reason in such situations, a simple hashCode comparison would be sufficient to reliably determine if our 2 BuiltList instances are made up of the same strings without having to iterate over them and do all those extra steps.

```
bool opreator==(other) {
    return other is BuiltList && other.hashCode == hashCode;
}
```

**Question:**  what am I missing, why the seemingly wasteful algorithm for comparison?

**Comment by davidmorgan** [comment]
Hi Daniel,

Thanks for the detailed issue.

Such an equals method will work most of the time, but it fails as soon as there are hash collisions. That is, if different lists end up with the same hash. Since we're using a 32 bit hash, the best case is a one in four billion chance of a collision. (If the hash function is perfect).

That sounds like a lot, but due to the birthday paradox you need to square root it for collections. So actually the number of hashes that will lead to a problem is only 2^16, i.e. 65536. This is a number that is easily small enough to turn up in practice.

An equals method is supposed to work 100% of the time :) ... if people want "probably equals", which would indeed be much faster, they can directly check the hashCode. This will probably happen as part of an algorithm that takes into account the probability of collisions and does check using "equals" when needed.

**Comment by daniel-v** [comment]
@davidmorgan, thanks for the answer. Since it's not really an issue, we can close this :)

TL;DR: hashCode match does not guarantee equality, hence the algorithm.

**Comment by davidmorgan** [comment]
It's interesting to think about adding a wide enough hash that it _can_ be used for equality, though. You could keep 128 bit hashes, then you should be okay with ~2^64 items. Plenty :)

I think we should have a use case for it first, though, i.e. an algorithm/system that's slow because of equality on lists and needs speeding up.

For UI code of the type I work on lists are usually limited in size to what you can display to a user, and the major cost is rendering rather than data processing. So, not much benefit to be had.
